### PR TITLE
Dylanra/clip based rag

### PR DESCRIPTION
**Pull request by dylanra-openai**
## Summary

Adding CLIP embedding RAG.

## Motivation

Showing users how to use multimodal RAG with direct image embeddings

---

## For new content

When contributing new content, read through our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md), and mark the following action items as completed:

- [x] I have added a new entry in [registry.yaml](https://github.com/openai/openai-cookbook/blob/main/registry.yaml) (and, optionally, in [authors.yaml](https://github.com/openai/openai-cookbook/blob/main/authors.yaml)) so that my content renders on the cookbook website.
- [x] I have conducted a self-review of my content based on the [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md#rubric):
  - [x] Relevance: This content is related to building with OpenAI technologies and is useful to others.
  - [x] Uniqueness: I have searched for related examples in the OpenAI Cookbook, and verified that my content offers new insights or unique information compared to existing documentation.
  - [x] Spelling and Grammar: I have checked for spelling or grammatical mistakes.
  - [x] Clarity: I have done a final read-through and verified that my submission is well-organized and easy to understand.
  - [x] Correctness: The information I include is correct and all of my code executes successfully.
  - [x] Completeness: I have explained everything fully, including all necessary references and citations.

We will rate each of these areas on a scale from 1 to 4, and will only accept contributions that score 3 or higher on all areas. Refer to our [contribution guidelines](https://github.com/openai/openai-cookbook/blob/main/CONTRIBUTING.md) for more details.

### CODE REVIEW

File: examples/custom_image_embedding_search.ipynb [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+        "\n",
+        "Building on work done in the [clothing matchmaker cookbook](https://cookbook.openai.com/examples/how_to_combine_gpt4v_with_rag_outfit_assistant), instead of using GPT-4V to caption an image and perform RAG on the caption, we instead embed the image directly as use that as the basis of our similarity search. The reason for this is in certain situations, there is information loss in converting an image to text as the text cannot fully capture the richness of the image. Therefore this direct image embedding may improve the most crucial part of RAG, the retrieval.\n",
+        "\n",
+        "An additional benefit of using CLIP-based embeddings is that we are also able to finetune CLIP to an enterprises' proprietary data or to update it with images it has not yet seen.\n",
+        "\n",
+        "The application area we will use to bring this technique to life is performing RAG over user inputted images of technology and searching our enterprise knowledge base for a matching image to provide helpful information for the end user."
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:24]: This context is useful but could be shortened and rephrased a bit – (sharing the ChatGPT prompt I use below)
```
Condense the following text, while keeping the word choice and tone the same.
```

File: examples/custom_image_embedding_search.ipynb [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+      "metadata": {
+        "id": "GgrlBLTpT0si"
+      },
+      "source": [
+        "Then let's import all the needed packages.\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 44,
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:66]: nit: Make sure to re-run your file in order so these are sequential and consecutive

File: examples/custom_image_embedding_search.ipynb [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+        "from torch.utils.data import DataLoader\n",
+        "import clip\n",
+        "import json\n",
+        "import matplotlib.pyplot as plt\n",
+        "import os\n",
+        "import base64\n",
+        "import numpy as np\n",
+        "import pickle\n",
+        "from typing import List, Union, Tuple\n",
+        "import faiss"
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:88]: These are lots os standard packages mixed with specific utils. Might want to split out the "meat" (openai, faiss, torch) from the standard (json, os, etc) and the specific/viz (matplotlib, tqdm, etc). And for the "meat" ones, would be useful to explain what you'll be using each one for, at a very high level.

File: examples/custom_image_embedding_search.ipynb [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": 45,
+      "metadata": {
+        "id": "_Ua9y98NRk70"
+      },
+      "outputs": [],
+      "source": [
+        "#load model on device\n",
+        "device = \"cpu\"\n",
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:109]: nit: comment on what this is (i'm assuming cpu vs gpu?)

File: examples/custom_image_embedding_search.ipynb [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+        "The application area we will use to bring this technique to life is performing RAG over user inputted images of technology and searching our enterprise knowledge base for a matching image to provide helpful information for the end user."
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "nbt3evfHUJTZ"
+      },
+      "source": [
+        "### Installations and imports\n",
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:33]: nit: use single # for top-level headers
(also see comment about explaining why you're install certain packages/imports below)

File: examples/custom_image_embedding_search.ipynb [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+        "import faiss"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "9fONcWxRqll8"
+      },
+      "source": [
+        "Now let's load the CLIP model."
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:97]: Add a section header. Would also be helpful to explain the steps you'll be taking throughout the cookbook at a high-level here (briefly) so the reader knows what's coming and what to expect.

File: examples/custom_image_embedding_search.ipynb [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 46,
+      "metadata": {
+        "id": "vE9i3zLuRk5c"
+      },
+      "outputs": [],
+      "source": [
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:133]: i'd split this up into separate code blocks, and put your longer comment explanations as text before. then you can give the tricky bits more space (and comments) e.g. get_features_from_image_path

File: examples/custom_image_embedding_search.ipynb [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
...1E4Opq6+Vt80+5KKOpwTywE9e1ZcbOoNeOuHr57VnPFMSWhKjAbRP0wzg4YOPTVvNPT1XCJLDbl9hOHd815HYnWiCqWivejAsHzwBgx0AgAEO79NnOvVMGsu3f+B0g4GRD/DgfOdCBNOtzYFPPERNvxoA6hyuEpAsLvzAGLgqLXyIDot5vSE7h1DQoO8AzzUloEQC2QxubUp3gP3T6tUxYbu/7GE27b17u3u9Nzzy/d+cw2cVsUOgM0IgRaUoTa7egVO0c8wKM5ruOg10vvYMiP63wUp/FMxYwA5BkJhvXYUr7eZyPsv/VBd/4gOGIn5n1Ju5mALU/RgyNNz/0i6ador1iB01JTFPf5StFHCC0IAY+pnkGkVwLlQAmdgigIWqV2qruqsiSs7yqyp2UsBmRL8W1IR63ZsnPmlHS6wGK0T925sa+G6nY+UIl6mkyWia61va5G1wD0kgVtOk6Az0GgEybZG1TrI03Pdonk+JlpmgfapsyQ0Be0gxYO3s7EfTkUrsmokaRbmw4yuNCDR1JmItoq0I6F4d7jGmaUkPp4cidVL7ZzbErAaa6MhyssO7+YaaQvI0npvLmpckEjdzGpck/j0EXFvR7Cqjl/ZrL5xIE+X/WfraE4/F92Gw1IwBkfNg3xU4Ul6tpLzPo5/xDl7C4X/qC1RPpalULJ4CU65aaXoaw7lnKS161olkCA0CsfUQA1NttG4xey3iRZdi2r6xwVIrvAjoNPtqkU/G719uXGdeQNUIQzzadDeCg/nFLo/N8beLR5hLNNAAXSO+U1hHMtkf7k6A75xSwOyZPcyuXTcdCy0+fPCRydHF4yQHKSMLMHZukiTSgnnm59HqfjmD/50nz5EHdTtW1bZdWhayakAagjHbXGGOM72Trsbmsn9XmZN7s1GkrbxoOdbDmDArOIOzQMpHUhHiZHxRbvpFrZmxNFs6oWg2KRiitcVloTowRCLOjTLwj9dSGaL7r+2FDiqltOwMohWLr9maMc/4eTZ/01yBmUu5KsjTmbWQEytoHAMgMPyt6c8RIP/INmwl5Fn70wqOvfFO1nl09jIcrUN0FJ9vWQ+vAh2DLfpVDYnRskT/zsWyUW16V1F4GdOSRgWMrh8NVZA+7AeK0XRhRvlSAACnSdoW6vqaIac+zQOFztJTNGHLgxKZxtL5Onlk5lsjkvxq18x9pPjQ/2vFm4K
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:414]: inquire*

Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:484]: nit: add conclusion paragraph

File: examples/custom_image_embedding_search.ipynb [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
...AAJicn8dJLL+lznnrqKZTLZT2p1QJOnz6N8fFxtLe3A7h020UphTvvvBOPPPIInnrqKfT39zuOL2bsDAwM4NVXX3UIRU888QSi0Sh27NixNhVZYVRrFxNeeeUVAHD0mUutXUwol8uYnp7e2H1l3UIIl4mf/OQnqrGxUT300EPqjTfeUF/84hdVU1OTI0LyUsfXvvY1tX//fnXs2DH17LPPqt27d6tEIqFGRkaUUkp96UtfUj09Peqpp55SL774ohoYGFADAwPrXOqVRyaTUS+//LJ6+eWXFQD1d3/3d+rll19WJ06cUEop9f3vf181NTWpRx99VB06dEh99KMfVf39/apQKOh73Hzzzeo973mPeuGFF9Qzzzyjtm7dqj75yU+uV5VWBJXaJZPJqL/8y79UBw8eVMeOHVO//vWv1Xvf+161detWVSwW9T0uxXa54447VCwWU/v371eDg4P6k8/n9TnVxs7s7KzauXOn2rNnj3rllVfUL3/5S9Xa2qruueee9ajSiqBauxw5ckR997vfVS+++KI6duyYevTRR9WmTZvU9ddfr+9xKbbLN77xDXXgwAF17NgxdejQIfWNb3xDeTwe9atf/UoptXH7yoYjfaWU+ud//mfV09OjfD6fev/736+ef/759S7SmuITn/iEam9vVz6fT3V2dqpPfOIT6siRI/p4oVBQf/7nf67i8bgKBoPq4x//uBocHFzHEq8Onn76aQVgwef2229XSp1ftvdXf/VXKplMqsbGRnXjjTeqw4cPO+4xPj6uPvnJT6pwOKyi0aj60z/9U5XJZNahNiuHSu2Sz+fVnj17VGtrq2poaFC9vb3qC1/4wgKh+VJsF1ObAFAPPvigPmcxY+f48ePqIx/5iAoEAiqRSKivfe1rqlQqrXFtVg7V2uXkyZPq+uuvV83NzaqxsVFt2bJFff3rX1dTU1OO+1xq7fJnf/Znqre3V/l8PtXa2qpuvPFGTfhKbdy+YrfWtbCwsLCwqBFsKJ++hYWFhYWFxfJhSd/CwsLCwqJGYEnfwsLCwsKiRmBJ38LCwsLCokZgSd/CwsLCwqJGYEnfwsLCwsKiRmBJ38LCwsLCokZgSd/CwsLCwqJGYEnfwsLCwsKiRmBJ38LCwsLCokZgSd/CwsLCwqJGYEnfwsLCwsKiRvD/AQGIBm//Wc2PAAAAAElFTkSuQmCC",
```
Comment (ibigio) [inline examples/custom_image_embedding_search.ipynb:197]: put image in images folder, add reference here (examples in [assistants cookbook](https://cookbook.openai.com/examples/assistants_api_overview_python))

**Review by ibigio — CHANGES_REQUESTED** [review]
| Criteria     | Description                                                                                         | Score |
| ------------ | --------------------------------------------------------------------------------------------------- | ----- |
| Relevance    | Is the content related to building with OpenAI technologies? Is it useful to others?                |    4   |
| Uniqueness   | Does the content offer new insights or unique information compared to existing documentation?       |     4  |
| Clarity      | Is the language easy to understand? Are things well-explained? Is the title clear?                  |     2  |
| Correctness  | Are the facts, code snippets, and examples correct and reliable? Does everything execute correctly? |    4   |
| Conciseness  | Is the content concise? Are all details necessary? Can it be made shorter?                          |     3  |
| Completeness | Is the content thorough and detailed? Are there things that weren’t explained fully?                |    4   |
| Grammar      | Are there grammatical or spelling errors present?                                                   |     3  |

Thanks for writing this up! The topic is super useful, excited to see share this. Left a few comments – generally this cookbook needs some work on structure and clarity (e.g. use of sections and headings, high-level explanations before jumping into details, and some splitting out / refactoring of code blocks).

Happy to take another look after a pass at these – this is super useful content!

**Review by ibigio — APPROVED** [review]
Thanks for incorporating the feedback! Left a few comments for minor typos, etc. Final note is to consider a title like "Multimodal RAG with CLIP Embeddings and GPT-4 Vision". Excited for this to ship!

File: examples/custom_image_embedding_search.ipynb [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "7hgrcVEl0Ma1"
+      },
+      "outputs": [],
+      "source": [
+        "#installations\n",
+        "%pip install clip\n",
```
Comment (Androbin) [inline examples/custom_image_embedding_search.ipynb:54]: Are you sure this is the right package?
https://pypi.org/project/clip/

**Pull request merged**